openapi_python_client/parser/properties/property.py

```diff
@@ -32,21 +33,39 @@ class Property:
     def __attrs_post_init__(self) -> None:
         object.__setattr__(self, "python_name", utils.to_valid_python_identifier(utils.snake_case(self.name)))
 
-    def get_type_string(self, no_optional: bool = False) -> str:
+    def get_base_type_string(self) -> str:
+        return self._type_string
+
+    def get_type_string(self, no_optional: bool = False, query_parameter: bool = False, json: bool = False) -> str:
         """
         Get a string representation of type that should be used when declaring this property
 
         Args:
             no_optional: Do not include Optional or Unset even if the value is optional (needed for isinstance checks)
+            query_parameter: True if the property's type is being used for a query parameter
+            json: True if the type refers to the property after JSON serialization
         """
-        type_string = self._type_string
+        if json:
+            type_string = self._json_type_string
+        else:
+            type_string = self.get_base_type_string()
+
         if no_optional:
-            return self._type_string
-        if self.nullable:
-            type_string = f"Optional[{type_string}]"
-        if not self.required:
-            type_string = f"Union[Unset, {type_string}]"
-        return type_string
+            return type_string
+        if self.required:
+            if self.nullable:
+                return f"Optional[{type_string}]"
```

Comment:
I think we're still not in alignment on design here. `nullable` should be the same as `not required` semantically when considering a query parameter. So `if required and nullable and query_parameter` it should be returning `f"Union[Unset, None, {type_string}]"`. I believe my previous suggestion to this (and the other, similar) code block makes this happen.

It's possible I'm missing or misunderstanding something about this, and if so please let me know. Otherwise I'll hold off additional review until you make that change.

Reply:
Hmm...if `required and nullable and query_parameter` are all `True`, wouldn't we want the type to be `Optional[{type_string}]`, *not* allowing `Unset`? Otherwise, if you allowed `Unset`, why would you make it "required"?

Or is it different for query parameters, where "required" would still allow `Unset` to pass through.

Reply:
It's different for query parameters. Basically we're not allowing `nullable` query parameters and instead treating `nullable` like `not required`. Additionally, we're allowing `None` to function just like `Unset`, but either way we're not passing through the parameter. I think having a different result for `nullable and required` might infer to the user that we can/will actually pass through a `None`, which we will not.

Reply:
I see. In that case, we should make the default `None` as well. I will change it. cc @bowenwr

Reply:
@dbanty I started implementing this approach, but it causes a bunch of other errors because in the templates, we are assuming that if something is `required and nullable`, then we only need to check for `None`, whereas with this change we now also need to check for `Unset`.

I think this is getting pretty complicated and hard to reason about, and I worry about future maintainability. Maybe we should just make the type `Optional` if it's either not required or nullable, and throw away `Unset` in the case of query parameters? cc @bowenwr

Reply:
In my original implementation of `Unset` I didn't apply it to query parameters. I'm personally in favor of that approach.

Reply:
If we have to pick one or the other I'd prefer `Unset` over `None` because it better indicates what we're doing with that parameter to the user. It also happens to be the behavior of the released version for non-required query params. What's your opinion @emann?